FLT/DivisionAlgebra/Finiteness.lean

```diff
@@ -118,19 +119,217 @@ lemma X_meets_kernel' {β : D_𝔸ˣ} (hβ : β ∈ ringHaarChar_ker D_𝔸) :
+  intro H x
+  obtain ⟨U, Uopen, Ueq⟩ := H (h.symm x)
+  exact ⟨U, Uopen, by simp [Ueq]⟩
+
+theorem XdiscreteY_if_HomoXdiscreteY {X X' Y : Type*} [TopologicalSpace Y] (f : X → Y)
+    (h : X ≃ X') : (∀ x' : X', ∃ U' : Set Y, IsOpen U' ∧ h '' (f ⁻¹' U') = {x'}) →
+    (∀ x : X, ∃ U : Set Y, IsOpen U ∧  f⁻¹' U = {x}) := by
+  intro H x
+  obtain ⟨U, Uopen, Ueq⟩ := H (h x)
+  refine ⟨U, Uopen, ?_⟩
+  simp_rw [(Equiv.eq_preimage_iff_image_eq h (f ⁻¹' U) {h x}).mpr Ueq,
+    Equiv.preimage_eq_iff_eq_image, Set.image_singleton]
+
+theorem DiscreteXinY_iff_DiscreteHomoXinY {X X' Y : Type*} [TopologicalSpace Y] (f : X → Y)
+    (h : X ≃ X') : (∀ x' : X', ∃ U' : Set Y, IsOpen U' ∧ h '' (f ⁻¹' U') = {x'}) ↔
+    (∀ x : X, ∃ U : Set Y, IsOpen U ∧  f⁻¹' U = {x}) :=
+  ⟨XdiscreteY_if_HomoXdiscreteY f h, HomoXdiscreteY_if_XdiscreteY f h⟩
+
+omit [MeasurableSpace (D ⊗[K] AdeleRing (𝓞 K) K)] [BorelSpace (D ⊗[K] AdeleRing (𝓞 K) K)] in
+theorem D_discrete_extracted (U : Set (Fin (Module.finrank K D) → AdeleRing (𝓞 K) K)) :
+    incl_Kn_𝔸Kn K D ⁻¹' U  = (D_iso K D) ''
+    (⇑(D𝔸_iso_top K D) ∘ (Algebra.TensorProduct.includeLeft : D →ₐ[K] D_𝔸) ⁻¹' U) := by
+  ext x
+  constructor
+  · intro hx
+    use (D_iso K D).symm x
+    unfold D𝔸_iso_top D𝔸_iso D_iso
+    simpa [← Algebra.algebraMap_eq_smul_one] using hx
+  · intro ⟨y, hy1, hy2⟩
+    have : (D𝔸_iso_top K D) ∘ (Algebra.TensorProduct.includeLeft : D →ₐ[K] D_𝔸) =
+        (incl_Kn_𝔸Kn K D) ∘ (D_iso K D) := by
+      ext d n
+      unfold incl_Kn_𝔸Kn D_iso D𝔸_iso_top D𝔸_iso
+      simp [← Algebra.algebraMap_eq_smul_one]
+    rw [this] at hy1
+    simpa [← hy2] using hy1
+
+omit [MeasurableSpace (D ⊗[K] AdeleRing (𝓞 K) K)] [BorelSpace (D ⊗[K] AdeleRing (𝓞 K) K)] in
+theorem D_discrete : ∀ x : D, ∃ U : Set D_𝔸,
+    IsOpen U ∧ (Algebra.TensorProduct.includeLeft : D →ₐ[K] D_𝔸) ⁻¹' U = {x} := by
+  apply XdiscreteY_if_XdiscreteHomeoY (Y' := ((Fin (Module.finrank K D) → AdeleRing (𝓞 K) K)))
+    (Algebra.TensorProduct.includeLeft : D →ₐ[K] D_𝔸) (D𝔸_iso_top K D)
+  apply XdiscreteY_if_HomoXdiscreteY (X' := Fin (Module.finrank K D) → K)
+    ((D𝔸_iso_top K D) ∘ (Algebra.TensorProduct.includeLeft : D →ₐ[K] D_𝔸)) (D_iso K D)
+  simpa [D_discrete_extracted] using Kn_discrete K D
+
+/-- The additive subgroup with carrier defined by Algebra.TensorProduct.includeLeft. -/
+local instance includeLeft_subgroup : AddSubgroup D_𝔸 := by
+  exact AddMonoidHom.range (G := D) (Algebra.TensorProduct.includeLeft : D →ₐ[K] D_𝔸)
+
+local instance : DiscreteTopology (includeLeft_subgroup K D).carrier := by
+  rw [includeLeft_subgroup]
+  apply (singletons_open_iff_discrete).mp
+  intro a
+  obtain ⟨a, a', ha⟩ := a
+  obtain ⟨U, hUopen, hUeq⟩ := (D_discrete K D) a'
+  refine isOpen_mk.mpr ⟨U, hUopen, Set.image_val_inj.mp ?_⟩
+  simp only [Subtype.image_preimage_coe, Set.image_singleton]
+  ext d
+  constructor
+  · intro hd
+    obtain ⟨hd1, hd2⟩ := hd
+    apply Set.mem_range.mp at hd1
```

Comment:
This is definitional so you don't need it if you're about to deconsrtuct `hd1` anyway.